feat(manager): add mise package manager (#29950)
renovatebot/renovate: lib/modules/manager/mise/readme.md

Keeps the [mise](https://mise.jdx.dev/configuration.html#mise-toml) `.mise.toml` file updated.

> `mise` supports versioning for many different tools (through it's [own plugins](https://github.com/jdx/mise/tree/main/src/plugins/core) and [asdf plugins](https://mise.jdx.dev/registry.html)).
> Therefore, specific tool support must be added one-by-one.
> 
> The following tools are currently supported:
> 
> <!-- Autogenerate in https://github.com/renovatebot/renovate -->
> <!-- Autogenerate end -->
> 
> <!-- prettier-ignore -->
> !!! note
>     Only the first version entry for each supported tool is managed, due to `.mise.toml` supporting fallback versions.

Reviewer: ```suggestion
Renovate's `mise` manager can version these tools:

<!-- Autogenerate in https://github.com/renovatebot/renovate -->
<!-- Autogenerate end -->

<!-- prettier-ignore -->
!!! warning
    Only the _first_ version entry for each supported tool is managed!
    This is due to `.mise.toml` supporting fallback versions.

To understand the versioning, Renovate re-uses:

- [mise's plugins](https://github.com/jdx/mise/tree/main/src/plugins/core)
- and [asdf's plugins](https://mise.jdx.dev/registry.html)

This also means that support for a new tool's versioning must be added one-by-one.
```

Changes:

- Put the important info for the reader up front, so the list of supported tools goes on top
- Bump the alert level of the admonition to a `warning` to grab attention better
- Make a list of the plugins Renovate re-uses
- Write short sentences
- Use simple words and phrases
- Split long sentences into two

Renovate can update the [mise](https://mise.jdx.dev/configuration.html#mise-toml) `.mise.toml` file.

Renovate's `mise` manager can version these tools:

<!-- Autogenerate in https://github.com/renovatebot/renovate -->
<!-- Autogenerate end -->

<!-- prettier-ignore -->
!!! warning
    Only the _first_ version entry for each supported tool is managed!

>     This is due to `.mise.toml` supporting fallback versions.

Reviewer: It feels we're missing a step in the explanation. Here's what I think when I read the current message:

1. Okay, so only the _first_ version for each tool is managed.
2. Looks like `.mise.toml` fallback versions cause this behavior.
3. But "why/how" does the fallback version "break" the way Renovate "sees/upgrades" versions?

Can you please improve the explanation, so the logical leap from "fallback version" to "broken Renovate" is clear? :wink:

Renovate can update the [mise](https://mise.jdx.dev/configuration.html#mise-toml) `.mise.toml` file.

Renovate's `mise` manager can version these tools:

<!-- Autogenerate in https://github.com/renovatebot/renovate -->
<!-- Autogenerate end -->

<!-- prettier-ignore -->
!!! warning
    Only the _first_ version entry for each supported tool is managed by Renovate. This limitation arises from the way `.mise.toml` is designed to handle fallback versions.
    Specifically, when multiple versions are specified for a tool, `.mise.toml` allows for fallback to secondary versions if the primary version is unavailable.
    Example: 'erlang = ["23.3", "22.0"]'
    Fallback versions provide a safety net in case the primary version is not accessible - the first version provided is the primary version and should be used where possible.

>     Renovate will only update the primary version - it is the users' responsibility to maintain functioning fallback versions.

Reviewer: Thank you for the extra explanation! :heart: The text is now too big for a admonition. So let's give the text some space, and nice headings. I also attempted another rewrite, to use simpler words and phrases. Feel free to correct any technical errors. :wink:

Do you think we should recommend users only use a _single_ version? Or will that cause more problems?

Here's my suggested text, for you to copy/paste/edit:

```markdown
### Renovate only updates primary versions

Renovate only updates the _first_ version it finds for each tool listed in the `mise.toml` file.
You must manually update any fallback versions!

For example if you had this snippet in your `mise.toml` file:

`'erlang = ["23.3", "22.0"]'`

Then Renovate would update the first, primary version: `"23.3"` but ignore the fallback `"22.0"` version.

#### Why can Renovate only update primary versions?

Usually the mise package manager can access the primary version, and skips any fallback versions.

But when the mise package manager accesses a fallback version, Renovate gets confused.

To be safe, Renovate only updates the _first_ version it sees.

## Renovate re-uses these plugins
```

Renovate can update the [mise](https://mise.jdx.dev/configuration.html#mise-toml) `.mise.toml` file.

Renovate's `mise` manager can version these tools:

<!-- Autogenerate in https://github.com/renovatebot/renovate -->
<!-- Autogenerate end -->

### Renovate only updates primary versions

Renovate's `mise` manager is designed to automatically update the _first_ (primary) version listed for each tool in the `mise.toml` file.

Secondary or fallback versions require manual updates.

#### Example

Given a `.mise.toml` entry like:

```toml
[tools]
erlang = ["23.3", "22.0"]
```

Renovate will update `"23.3"` (the primary version) but will not touch `"22.0"` (the fallback version).

#### Why can Renovate only update primary versions?

To maintain consistency and reliability, Renovate opts to only manage the _first_ listed version.

- Fallback versions can often be older versions of a tool that are known to work and are there as a backup.

This follows the same workflow that Renovate's `asdf` manager uses.

> ### Plugin/Tool support

Reviewer: ```suggestion
### Plugin/tool support
```

Renovate uses:

- [mise's plugins](https://github.com/jdx/mise/tree/main/src/plugins/core)
- [asdf's plugins](https://mise.jdx.dev/registry.html)

to understand and manage tool versioning.

Support for new tools/plugins needs to be _manually_ added to Renovate's logic.

#### Adding new tool support

There are 2 ways to integrate versioning for a new tool:

- Renovate's `mise` manager: ensure upstream `mise` supports the tool, then add support to the `mise` manager in Renovate
- Renovate's `asdf` manager: improve the `asdf` manager in Renovate, which automatically extends support to `mise`

> If `mise` adds support for more tools via it's own [core plugins](https://mise.jdx.dev/plugins.html#core-plugins), you can create a PR to extend Renovate's `mise` manager to add support for the new tooling.

Reviewer: ```suggestion
If `mise` adds support for more tools via its own [core plugins](https://mise.jdx.dev/plugins.html#core-plugins), you can create a PR to extend Renovate's `mise` manager to add support for the new tooling.
```

> - You may be able to add support for new tooling upstream in the core plugins - create an issue and see if the community agrees whether it belongs there, or if it would be better as an `asdf-` plugin.

Reviewer: ```suggestion
You may be able to add support for new tooling upstream in the core plugins - create an issue and see if the community agrees whether it belongs there, or if it would be better as an `asdf-` plugin.
```

Let's drop the bullet point here, and keep it as plain text.

<!-- Autogenerate end -->

<!-- prettier-ignore -->
### Renovate only updates primary versions

Renovate's `mise` manager is designed to automatically update the _first_ (primary) version listed for each tool in the `mise.toml` file.

Secondary or fallback versions require manual updates.

#### Example

Given a `.mise.toml` entry like:

```toml
[tools]
erlang = ["23.3", "22.0"]
```

Renovate will update `"23.3"` (the primary version) but will not touch `"22.0"` (the fallback version).

#### Why can Renovate only update primary versions?

To maintain consistency and reliability, Renovate opts to only manage the _first_ listed version.

- Fallback versions can often be older versions of a tool that are known to work and are there as a backup.

This follows the same workflow that Renovate's `asdf` manager uses.

### Plugin/Tool support

Renovate leverages:

- [mise's plugins](https://github.com/jdx/mise/tree/main/src/plugins/core)
- [asdf's plugins](https://mise.jdx.dev/registry.html)

for understanding and managing tool versioning.

Support for new tools/plugins needs to be _manually_ added to Renovate's logic.

#### Adding new tool support

> Adding support for a new tool involves either contributing to Mise Plugins or extending Renovate's asdf manager, which indirectly benefits mise management.

Author: Rewritten it a lil' bit - mise uses asdf's plugin ecosystem under the hood, so adding support for `asdf-x` plugins is arguably one of the easiest ways to add support for tooling into `mise`.

`mise` does have core plugins which "are new and will improve over time. ([ref](https://mise.jdx.dev/plugins.html#core-plugins)), but I'm not sure about user contribution to them (i.e. it looks like it's only for main language versions rather than tools generically)

> Adding support for a new tool involves either contributing to Mise Plugins or extending Renovate's asdf manager, which indirectly benefits mise management.

Reviewer: I'm happy with your text. I don't know enough to decide which method is best, so I'll let you judge what's best. :smile: 

Readers can now use the information in the docs to decide how they want to add support for new tools to Renovate.
